mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,37 +458,111 @@ impl EthCoin {
+        let decoded = decode_contract_call(approve_func, taker_payment.unsigned().data())?;
+        let id = match decoded.first() {
+            Some(Token::FixedBytes(bytes)) => bytes,
+            invalid_token => {
+                return MmError::err(FindPaymentSpendError::InvalidData(format!(
+                    "Expected Token::FixedBytes, got {:?}",
+                    invalid_token
+                )))
+            },
+        };
+        // loop to find maker's spendTakerPayment transaction
         loop {
-            let taker_status = self
-                .payment_status_v2(
-                    taker_swap_v2_contract,
-                    decoded[0].clone(), // id from spendTakerPayment
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
+            }
+
+            let current_block = match self.current_block().compat().await {
+                Ok(b) => b,
+                Err(e) => {
+                    error!("Error getting block number: {}", e);
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // get all logged TakerPaymentSpent events from `from_block` till current block
+            let events = match self
+                .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block, current_block)
+                .await
+            {
+                Ok(events) => events,
+                Err(e) => {
+                    error!(
+                        "Error getting TakerPaymentSpent events from {} block: {}",
+                        from_block, e
+                    );
+                    Timer::sleep(5.).await;
+                    continue;
+                },
+            };
+
+            // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+            let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
+
+            if let Some(event) = found_event {
+                if let Some(tx_hash) = event.transaction_hash {
+                    let transaction = match self.transaction(TransactionId::Hash(tx_hash)).await {
+                        Ok(Some(t)) => t,
+                        Ok(None) => {
+                            info!("Tx {} not found yet", tx_hash);
+                            Timer::sleep(check_every).await;
+                            continue;
```

Comment:
now that we have the tx_hash, we shouldn't keep looping right?

Reply:
hmm yes
```rust
            if let Some(event) = found_event {
                if let Some(tx_hash) = event.transaction_hash {
                    let transaction = match self.transaction(TransactionId::Hash(tx_hash)).await {
                        Ok(Some(t)) => t,
                        Ok(None) => {
                            info!("Tx {} not found yet", tx_hash);
                            Timer::sleep(check_every).await;
                            continue;
                        },
                        Err(e) => {
                            error!("Get tx {} error: {}", tx_hash, e);
                            Timer::sleep(check_every).await;
                            continue;
                        },
                    };
                    let result = signed_tx_from_web3_tx(transaction).map_err(FindPaymentSpendError::Internal)?;
                    return Ok(result);
                }
            }

            Timer::sleep(5.).await;
        }
```
once we got `Some(tx_hash)` it means event we found had necessary swap id.
But if we got None or error in next step after `Some(tx_hash)`, we dont have to repeat the whole process from the beginning. 
Next loop cycle we can start right from requesting transaction from `tx_hash`. I will work on it.

Reply:
Done https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/6c8b2c132cffd36a08f55f655f0eacfffa205882

Reply:
we should abort  if we got error as this can lead to a potential infinite loop. `self.transaction` return error only if the tx data can't be deserialize which means bad tx data so we should break

Reply:
> we should abort if we got error as this can lead to a potential infinite loop.

There is no infinite loop, we have `if now > wait_until` check
```rust
        loop {
            let now = now_sec();
            if now > wait_until {
                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
            }
```

> `self.transaction` return error only if the tx data can't be deserialize which means bad tx data so we should break

no, it has `try_rpc_send` function, it could return `web3::Error`
![image](https://github.com/user-attachments/assets/5675cfab-d06c-4112-8429-5277117f7788)

Reply:
ok.. then you still handle the error

Reply:
> ok.. then you should still handle the error

I think you confused `transaction` and `signed_tx_from_web3_tx` functions
https://github.com/KomodoPlatform/komodo-defi-framework/blob/f437dde06e77774407afc71829c29961faa3366a/mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs#L528-L541

`transaction` uses web3 calls.
`signed_tx_from_web3_tx` is one which doesnt have web3 requests and tries to build ethcore transaction from web3 transaction. If it fails then it is useless to repeat it, we can return error instantly and we do it
```
let transaction = signed_tx_from_web3_tx(t).map_err(FindPaymentSpendError::Internal)?;
```